proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,513 @@
+
+
+### Accessing Enclosing Self from Wrapper Types
+
+There's currently no public feature that allows a wrapper to access its enclosing `Self` type:
+
+```swift
+@propertyWrapper
+struct Mirror<EnclosingSelf, Value, Path>
+  where Path : KeyPath<EnclosingSelf, Value> { 
+  
+  let keyPath: Path 
+
+
+  init(of keyPath: Path) { ... }
+  
+}
+
+
+struct Point {
+
+  private var _vector: SIMD2<Double>
+  
+  
+  init(x: Double, y: Double) {
+    self._vector = SIMD2(x: x, y: y)
+  }
+  
+  
+  @Mirror(of: \._vector.x)
+  var x
+  
+  @Mirror(of: \._vector.y)
+  var y
+  
+} ❌
+// In the above use, we'd access the enclosing
+// type's '_vector' property through the provided
+// keyPath. However, today that's invalid.
+```
+
+Furthermore, extending this feature's availability to function and closure parameters be really powerful:
+
+```swift
+func valueAndIdPair<Value>(
+  @Mirror of property: Value
+) -> (value: Value, id: Int) {
+  (value: property, id: $property.keyPath.hashValue)
+}
+```
+
+It's important to note that allowing use of such a feature in function parameters would entail some limitations. For example, a parameter marked with a wrapper type referencing enclosing `self` would be an error for a non-instance method, as the accessors for the `wrappedValue` and `projectedValue` need an enclosing `self` instance.
+
+### Add Wrapper Types in the Standard Library
+
+Adding wrapper types to the Standard Library has been discussed for types [such as `@Atomic`] and [`@Weak`](https://forums.swift.org/t/should-weak-be-a-type/34032), which would facilitate certain APIs. Another interesting Standard Library wrapper type could be `@UnsafePointer`, which would be quite useful, as access of the `pointee` property is quite common:
+
+```swift
+let myPointer: UnsafePointer<UInt8> = ...
+
+myPointer.pointee 
+//        ^~~~~~~ 
+// This is the accessor pattern property 
+// wrappers were devised to tackle.
+```
+
+Instead of writing the above, in the future one might be able to write this:
+
+```swift
+let myInt = 0
+
+withUnsafePointer(to: ...) { @UnsafePointer value in
+
+  print(value) // 0
+  
+  $value.withMemoryRebound(to: UInt64.self {
+    ... 
+  }
+  
+}
+```
+
+As a result, unsafe code is not dominated by visually displeasing accesses to `pointee` members; rather, more natural and clear code is enabled. 
+
+What's more, a `@Lazy` type could be added so as to alleviate the need for custom behavior built into the compiler. Instead, `@Lazy` would act as more comprehensible, easy to maintain type, that would also allow for a more streamlined way of resetting its storage:
```

Comment:
I think if we are going to include this, we should use an example that uses `@Lazy` on a parameter. Otherwise, it seems strange that this is included in the "Property Wrapper Parameters" proposal

Reply:
I've given this some thought and I wanted to get your opinion. 

You proposed adding `@Lazy` as a replacement for `lazy` and `@autoclosure`. That's an amazing idea, but really difficult to implement if not impossible in the current language semantics. It's difficult to implement `@Lazy` as a replacement for `@autoclosure`, as it requires an immutable getter – which `@Lazy` lacks by design. There are also various other hurtles that `@Lazy` would have to overcome, which would make justifying the implementation effort difficult. 

Therefore, that leaves us with `lazy`, which I used in the examples. I do think, though, that you're right; we should end the proposal with an example that utilizes the feature we actually propose. Thus, I think we should look for other wrapper types that fit the criteria of being acceptable for the standard library and that utilize property wrappers in function or closure parameters.

Thoughts?

Reply:
You're right that if we want to use `@Lazy` on a parameter, it can't have a `mutating` getter. If we want to use this example, we'd have to implement `Lazy` differently than in the original property wrapper proposal. Using reference types is the way to do this, either by making `Lazy` itself a reference type, or making it store an instance of some reference type that represents the initialization state.

Reply:
Well, yeah there are various ways of achieving reference-semantics; what I’m skeptical about  is whether it’s worth going through a lot of trouble for a simple ‘@Lazy’ wrapper instead of finding a better wrapper for the job. Another wrapper type would avoid complicated discussion about what will just be a future direction. Namely, I expect review questions asking about aspects like the following:

1. Lazy properties currently exhibit value semantics; how will a single reference-semantics ‘@Lazy’ wrapper handle that?

2. Won’t ‘@autoclosure’ still be required?

3. Why spend so much resources on syntactic sugar?

Reply:
That's fair - it's not worth raising a bunch of questions just for a future direction. I think we should find another example or just remove the section. Somebody on the pitch thread had suggested creating a "manifesto" document for all of the future directions for property wrappers that have been floating around, and we could put all the work you've done for this section in there so it doesn't get lost!

Reply:
Yeah, I read that post and I think I this is a _great_ idea! There’s so much potential with property wrappers that we just can’t fit into a future-directions section. As for the wrapper to be used here, any suggestion is welcome; I’ll also be searching for a good wrapper type in personal and public projects.